### PR TITLE
C/C++/... stubs documentation

### PR DESCRIPTION
**Pull request by diml**
This PR adds a page to the manual explaining how to use foreign code with Dune. In particular, it covers how to integrate a foreign build system in a Dune build, which is not obvious.

### CODE REVIEW

File: doc/foreign-code.rst [hunk 1]
```diff
@@ -0,0 +1,64 @@
+******************************
+Dealing with foreign libraries
+******************************
+
+The OCaml programming language allows to interface libraries written
+in foreign languages such as C. This section explains how to do this
+with Dune. Note that it does not cover how to write the C stubs
+themselves, this is covered by the OCaml manual: LINK TO MANUAL.
```
Comment (emillon) [inline doc/foreign-code.rst]: That would be https://caml.inria.fr/pub/docs/manual-ocaml/intfc.html

File: doc/foreign-code.rst [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+files written in foreign languages. This support is suitable for most
+OCaml projects containing C stubs, but is too limited for building
+complex libraries written in C or other languages. For such cases,
+Dune allows to integrate a foreign build system into a normal Dune
+build.
+
+Adding C/C++ stubs to an OCaml library
+======================================
+
+To add C stubs to an OCaml library, simply list the C files with the
```
Comment (emillon) [inline doc/foreign-code.rst]: I think that should be `without`.

File: doc/foreign-code.rst (hunk 1 above)
Comment (diml) [reply doc/foreign-code.rst]: Fixed. I wrote this documentation in the plane and their proxy was blocking caml.inria.fr, can you believe it...

File: doc/foreign-code.rst (hunk 2 above)
Comment (diml) [reply doc/foreign-code.rst]: Fixed

**Comment by diml** [comment]
Yh, for ctypes it would be best to add a ctypes stanza to hide all the boilerplate.

**Comment by diml** [comment]
I finished this page.

File: doc/foreign-code.rst [hunk 3]
```diff
@@ -0,0 +1,150 @@
+******************************
+Dealing with foreign libraries
+******************************
+
+The OCaml programming language allows to interface libraries written
+in foreign languages such as C. This section explains how to do this
+with Dune. Note that it does not cover how to write the C stubs
+themselves, this is covered by the OCaml manual:
```
Comment (emillon) [inline doc/foreign-code.rst]: I think that this should be a link. The syntax is:

    `link test <http://example.org/>`_

File: doc/foreign-code.rst [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+https://caml.inria.fr/pub/docs/manual-ocaml/intfc.html.
+
+More precisely, this section covers:
+- how to add C/C++ stubs to an OCaml library
+- how to pass specific compilaction flags for compiling the stubs
```
Comment (emillon) [inline doc/foreign-code.rst]: typo: compilation

File: doc/foreign-code.rst [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+these are made visible to users of the library via the include search
+path.
+
+.. _foreign-sandboxing:
+
+Foreign build sandboxing
+========================
+
+When the build of a C library is too complicated to express in the
+Dune language, it is possible to simply *sandbox* an foreign
```
Comment (emillon) [inline doc/foreign-code.rst]: typo: a foreign build

File: doc/foreign-code.rst [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+                    (copy libfoo/libfoo.a libfoo_stubs.a)
+                    (copy libfoo/libfoo.so dllfoo_stubs.so)))
+
+Note that the rule copies the files to ``libfoo_stubs.a`` and
+``dllfoo_stubs.so``. It is important that the files produced are
+named ``lib<ocaml-lib-name>_stubs.a`` and
+``dll<ocaml-lib-name>_stubs.so``.
+
+The last step is to attach these archives to an OCaml library as
+follow:
```
Comment (emillon) [inline doc/foreign-code.rst]: typo: as follows

File: doc/foreign-code.rst [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+.. code:: scheme
+
+          (library
+           (name foo)
+           (self_build_stubs_archive foo))
+
+Then, whenever you use the ``foo`` library, you will also be able to
+use C functions from ``libfoo``. Note that the argument of
+``self_build_stubs_archive`` has to be the name of the library. If you
+put anything else it will be ignored, this is a bug in Dune.
```
Comment (emillon) [inline doc/foreign-code.rst]: I'd say "This is a known bug in dune". Maybe we can open an issue and link it there?

File: doc/foreign-code.rst (hunk 3 above)
Comment (diml) [reply doc/foreign-code.rst]: thanks, I updated the link

File: doc/foreign-code.rst (hunk 7 above)
Comment (diml) [reply doc/foreign-code.rst]: indeed

Comment (diml) [reply doc/foreign-code.rst]: done

**Pull request merged**